src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+            u1
+        ))
+        u0
+    ))
+)
+
+;; Add a PoX address to a given sequence of reward cycle lists.
+;; A PoX address can be added to at most 12 consecutive cycles.
+;; No checking is done.
+;; Returns the number of reward cycles added
+(define-private (add-pox-addr-to-reward-cycles (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+                                               (first-reward-cycle uint)
+                                               (num-cycles uint)
+                                               (amount-ustx uint))
+    (begin
+        ;; For safety, add up the number of times (add-principal-to-ith-reward-cycle) returns 1.
+        ;; It _should_ be equal to num-cycles.
+        (asserts! (is-eq num-cycles (fold +
+           (map add-pox-addr-to-ith-reward-cycle (list
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u0 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u1 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u2 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u3 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u4 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u5 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u6 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u7 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u8 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u9 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u10 }
+              { pox-addr: pox-addr, first-reward-cycle: first-reward-cycle, num-cycles: num-cycles, amount-ustx: amount-ustx, i: u11 }))
+           u0))
+           (err ERR-STACKING-UNREACHABLE))
+
+        ;; mark address in use over this range
+        (map-set pox-addr-reward-cycles
+            { pox-addr: pox-addr }
+            { first-reward-cycle: first-reward-cycle, num-cycles: num-cycles }
+        )
+
+        (ok true)
+    )
+)
+
+;; What is the minimum number of uSTX to be stacked in the given reward cycle?
+;; Used internally by the Stacks node, and visible publicly.
+(define-read-only (get-stacking-minimum (reward-cycle uint))
```

Comment:
For the >25% case, the limit should be scaled by participating stx, not the total liquid amount:

```scheme
(define-read-only (get-stacking-minimum (reward-cycle uint))
    (let ((ustx-stacked-so-far (get-total-ustx-stacked reward-cycle)))
      (let
        ((unrounded-limit
          (if (< ustx-stacked-so-far (/ total-liquid-ustx u4))
            ;; less than 25% of all liquid STX are stacked, so threshold is bound
            (/ total-liquid-ustx STACKING-THRESHOLD-25)
            ;; greater than 25% of liquid STX are stacked, so threshold is dynamic
            (/ ustx-stacked-so-far STACKING-THRESHOLD-100))))
        ;; round limit to nearest 10k
        (if (eq? 0 (mod unrounded-limit 10000))
            ;; already at nearest 10k
            unrounded-limit
            ;; round to next highest 10k
            (* (+ (/ unrounded-limit 10000) 1) 10000)))))
```

Reply:
Sure, will fix.  But, I'm curious why the rounding factor is 10k?  Is the reason to ensure that a Stacker will be included in the next reward cycle, despite tokens getting unlocked through mining?  Because if so, this value is kinda high.

If the coinbase is 500 STX, then in a fully-mined reward cycle where each Bitcoin block produces one Stacks block and there are no Stacks forks, then at the end of the reward cycle, the total liquid supply will have increased by 500,000 STX.  This means that each reward slot must be acquired for 100 more STX in the next reward cycle (this is 500,000 STX shared across 5,000 reward addresses).  If a user is going to Stack for multiple reward cycles, then they'll need to lock up an additional amount of STX _a priori_ to ensure that they meet the minimum Stacking threshold in all of their reward cycles.  But, if a user can only Stack for 12 reward cycles at a time, then the maximum additional _a priori_ amount to commit is only 1,200 STX, not 10,000 STX.  Am I missing something?

Reply:
It's just a round-ish constant, we can choose a different one. The goal of rounding is to satisfy the "tiering" discussion from (https://github.com/blockstack/stacks-blockchain/issues/1837) -- if we have a dynamic limit for Stacking, then setting a round limit makes deciding about how much to Stack much easier (i.e., your much less likely to accidentally lockup too little).

Reply:
I thought we had decided long, long ago that the only tiers we would have would be the the 25%-participation and 100%-partition tiers?

> your much less likely to accidentally lockup too little

I thought this problem was exactly the problem that delegation solved?  You and everyone else join your STX and lock into the highest tier's minimum stacking requirement.  Assuming the risk of using a delegate is zero, you're incentivized to band together with as many other people as possible, because it minimizes the opportunity cost of your lock-up.  You can get the delegate risk close to zero by using a multisig Bitcoin address, in which you and a majority (or even 100%) of the Stackers are signatories.

If this weren't the case, then tiering itself would create a game of "chicken" where everyone is incentivized to wait until the last possible moment to Stack their STX, since it minimizes the opportunity cost of locking up their tokens (while the upside is the same regardless of what tier you land in).  Moreover, it would encourage miners to offer off-chain services where individuals would pay off-chain to get get their STX locked into the lowest tier (and miners would exclude everyone else).

Reply:
> I thought we had decided long, long ago that the only tiers we would have would be the the 25%-participation and 100%-partition tiers?

Not that I'm aware of. See: https://github.com/blockstack/stacks-blockchain/blob/master/sip/sip-007-stacking-consensus.md#adjusting-reward-threshold-based-on-participation.

> If this weren't the case, then tiering itself would create a game of "chicken" where everyone is incentivized to wait until the last possible moment to Stack their STX, since it minimizes the opportunity cost of locking up their tokens (while the upside is the same regardless of what tier you land in). Moreover, it would encourage miners to offer off-chain services where individuals would pay off-chain to get their STX locked into the lowest tier, and miners would try to prevent people who didn't pay up from Stacking in the lowest tier.

There is just one limit for each reward cycle, regardless of when you lockup your STX, so there's no game of chicken, except that you may receive more information about what the minimum amount to lockup is (also note: if I lockup _more_ than the minimum `T`, then I get `floor(X/T)` reward addresses, so it's not like there's an enormous incentive to try and only lockup the minimum amount).

> I thought this problem was exactly the problem that delegation solved? You and everyone else join your STX and lock into the highest tier's minimum stacking requirement. Assuming the risk of using a delegate is zero, you're incentivized to band together with as many other people as possible, because it minimizes the opportunity cost of your lock-up.

Delegation can help in a lot of ways, but the limit scaling is done to reduce the number of "burned" slots when Stacking participation is not 100%, which delegation does not directly help with.

Reply:
> Not that I'm aware of. See: https://github.com/blockstack/stacks-blockchain/blob/master/sip/sip-007-stacking-consensus.md#adjusting-reward-threshold-based-on-participation.

I don't think we have the same definition of "tier"?  The text says `However, if participation is lower than 100%, the reward pool could admit lower STX holders. The Stacking protocol specifies 2 operating levels: ...`.  Doesn't this mean that there are two reward tiers?  i.e. one reward tier where the stacking minimum 1/20,000th of the liquid supply, and another tier where the stacking minimum is 1/5,000th of the liquid supply?  Are we now saying that there are more such tiers, and/or that the stacking minimum can now be a dynamic quantity?

> There is just one limit for each reward cycle, regardless of when you lockup your STX, so there's no game of chicken, except that you may receive more information about what the minimum amount to lockup is (also note: if I lockup more than the minimum T, then I get floor(X/T) reward addresses, so it's not like there's an enormous incentive to try and only lockup the minimum amount).

Those reward slots have the same absolute worth, regardless of how many people Stack.  As a Stacker, I want to acquire as many such reward slots as I can manage with my STX.  Stacking doesn't guarantee a reward address by itself, and Stacking without knowing how many people will Stack in my target reward cycle is risky if I can be priced out of a tier.  If I Stack earlier, and the Stacking minimum gets bumped up beyond the amount that I've stacked, then my STX would remain locked but I receive no reward.  _I want to avoid this if at all possible_, since I get no upside but I still pay the opportunity cost of locking.  If I wait to Stack later -- i.e. I consider Stacking at the very end of the reward cycle, after everyone else has Stacked -- then by the time I make my decision to Stack or not, I would either (1) _already_ be priced out of any reward addresses, but by waiting I at least still have my liquid STX (so, no opportunity cost); or (2) be in a position to Stack _and_ get a reward address, since if I get the last transaction in the reward cycle (or close enough such that this is unlikely), no one can price me out after I lock my STX.

All I was saying is that this dilemma of whether or not to Stack, and when to Stack, goes away if you're guaranteed to participate in the highest-priced tier (which I think you were pointing out?).  I was saying that delegation makes acquiring reward addresses with the highest minimum Stacking value imminently achievable for most participants.

> Delegation can help in a lot of ways, but the limit scaling is done to reduce the number of "burned" slots when Stacking participation is not 100%, which delegation does not directly help with.

The problem of "your much less likely to accidentally lockup too little" exists _because_ of limit scaling, I thought?  If we didn't do any scaling, there's no way for you to get priced out of a reward address after you lock your tokens -- you wouldn't have been able to lock them in the first place?

Reply:
>  The text says However, if participation is lower than 100%, the reward pool could admit lower STX holders. The Stacking protocol specifies 2 operating levels: ...

Immediately following that is a description of the operation:

```
If between 0.25 * STX_LIQUID_SUPPLY and 1.0 * STX_LIQUID_SUPPLY STX participate
in a reward cycle, the reward threshold is optimized in order to maximize the number of
slots that are filled. That is, the minimum threshold T for participation will be roughly
1/5,000th of the participating STX (adjusted in increments of 10,000 STX). 
```

Reply:
> That is, the minimum threshold T for participation will be roughly 1/5,000th of the **participating** STX (adjusted in increments of 10,000 STX).

Emphasis mine.

I feel foolish -- I wish I had comprehended this sooner, because this _significantly_ changes my perception of Stacking.  I would have spoken up _against_ this provision much earlier in the SIP-007 design process, because I think the possibility of Stacking too few tokens creates a bunch of perverse incentives for Stackers, their delegates, and miners.  Not only are Stackers and their delegates now encouraged to try and Stack as late as possible (so they don't fall into the trap of locking up their STX while also failing to qualify for a reward slot), but also they're encouraged to bribe miners to do things like delay, reorder, or block other peoples' Stacking transactions in order to avoid getting priced out.

I think, fundamentally, we need to make it so people are incentivized to Stack honestly.  There should be no downside risk for trying to Stack, except for paying the associated transaction fee.  People should be able to Stack their STX at any point in time, and if they get priced out after subsequent Stacking transactions for the next reward cycle, then their STX need to be returned to them once the next reward cycle begins.  This I don't think should be too bad to do lazily in the current implementation -- we'd just record in the account struct what the Stacker thought the minimum Stacking threshold was when they Stacked, and if the true minimum Stacking threshold was higher, we'd ignore the lock on balance queries in the next reward cycle and remove the lock on the next spend attempt in a subsequent reward cycle.

Reply:
So, summarizing, here's what I'm proposing I do to resolve this thread:

* Re-structure `(get-stacking-minimum ..)` as @kantai proposed.
* Make it so that Stackers who stacked in this reward cycle, but did not meet the minimum threshold in the next reward cycle, are automatically unlocked.

Sounds good?

Reply:
Sounds good to me!

Reply:
There is some confusion on my end, I think https://github.com/blockstack/stacks-blockchain/pull/1790#discussion_r479605225 superseded the consensus reached in this convo, but could be missing something.
Is this concern being addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/e22613ea957d0db50ef761131b67e29c77ffa6ab?